### ISSUE TITLE
Prometheus Operator GA / v1.0

### ISSUE THREAD

**Issue by sichvoge** [prometheus-operator/prometheus-operator#3932]
First of all. Thanks to everyone for their hard work on the Operator. 

I am wondering if there is any desire to move the Prometheus Operator to GA or v1.0 in the near future? Right now it's still labeled as "beta" but I feel that it is being used by a lot of people already and already went through so many cycles that it should be "beta" anymore. Any insights would be much appreciated.

**Comment by coderanger** [comment]
Beta has specific meaning in Kubernetes parlance. It means that the feature or subsystem (or in this case, project) is stable enough for general use but there is some expectation that backwards incompatible data changes are likely before GA. It doesn't imply that the project is unstable or that you shouldn't use it :)

**Comment by lilic** [comment]
cc @prometheus-operator/prometheus-operator-reviewers @prometheus-operator/prometheus-operator-triage please have a look. Thanks, Christian for opening this! 

I remember us having some checklist of features we wanted before making it go v1.x but I honestly forget what they were, so this might be good place to collect those.

**Comment by paulfantom** [comment]
I am aware that some of our APIs are stable already, but making project GA has some marketing considerations and it means most of the project we ship is (or should be) stable, not only our APIs. We recently introduced AlertmanagerConfig CRD and there is a lot of confusion around this, so we should probably try to graduate it to beta or stable before going GA. Additionally, we have quite a lot of configuration options that are marked as deprecated, and going GA seems like a good time to finally remove them (and possibly promote APIs to the next versions).

Also just wanted to point out that initial prometheus-operator design goals changed a bit during the project lifetime. So even if we had a list of things that needed to be done before having GA that list is probably quite different from what we have in mind now (or not :smile:). I think we should come up with a new list of priorities (maybe we need some sort of a roadmap?) and follow it.

---
Here is my braindump of what I would see for 1.0 GA:

Some examples of things to deprecate in APIs (even if we need the next API versions):
- `ArbitraryFSAccessThroughSMsConfig` as it is a workaround for accessing `bearerTokenFile` and used only in ServiceMonitor
- `StorageSpec.disableMountSubPath`
- `tag`, `sha`, `baseImage` in AlertmanagerSpec, PrometheusSpec, and ThanosSpec

Some things to consider for GA:
- [generic ScrapeConfig CRD](https://github.com/prometheus-operator/prometheus-operator/issues/2787) (base for ScrapeMonitor, PodMonitor, OtherMonitor)
- better sharding mechanism (hashmod works, but could be better ;) )
- [`.Status` field in CRs](https://github.com/prometheus-operator/prometheus-operator/issues/3335)
- [zero-downtime StatefulSet recreation](https://github.com/prometheus-operator/prometheus-operator/issues/3875)

**Comment by simonpasquier** [comment]
IMHO there are different levels of stability guarantees. On top of my head: 
* Custom resource definitions. For better or worse, most of them are v1 so they can already be considered "stable". For instance, getting rid of `ArbitraryFSAccessThroughSMsConfig` can't happen without bumping the version of the ServiceMonitor CRD. The only exception as @paulfantom mentioned is the AlertmanagerConfig CRD which is v1alpha1 (use it at your own risk).
* Supported Kubernetes versions.
* Supported versions of operands.
* The operator's flags.
 
My assumption is that users care about the stability of the CRDs, not so much about the rest.

**Comment by paulfantom** [comment]
Crazy thought, but completely serious - what if we ditch semver and follow a calendar release model (releases tagged with a date, ex. `v2021.06`), while still following CRD versioning we have right now? This would have the benefit of showing where the stability of the project lies (CRDs) and what is rolling to satisfy those APIs. 

Additionally, it would have a positive side-effect of quickly showing how old version of the operator someone is using :smile:

**Comment by paulfantom** [comment]
It would be good to finalize #3439 too.

**Comment by weyert** [comment]
Interesting, yes, would be nice to have some sort of matrix were each feature is listed and the level of stability :) Currently, it sounds like its better to avoid this useful operator due to its beta status.

**Comment by paulfantom** [comment]
> would be nice to have some sort of matrix where each feature is listed and the level of stability 

That is listed in API definition (version field) and particular field descriptions. But I agree it could be more visible.

**Comment by weyert** [comment]
@paulfantom Sorry, where can I find this API definition?

**Comment by paulfantom** [comment]
@weyert We are following kubernetes API versioning schema practices. This means each CRD has a version field saying how stable this particular API is. So:

- AlertmanagerConfig is v1alpha1 and subject to changes
- Alertmanager, PodMonitor, Probe, Prometheus, PrometheusRules, ServiceMonitor, and ThanosRuler are v1 and thus considered stable

Data is available in generated CRD files as well as in golang code in https://github.com/prometheus-operator/prometheus-operator/tree/master/pkg/apis/monitoring

**Comment by s-urbaniak** [comment]
I recall early discussion where we wanted to finalize support of status subresource before GA'ing prometheus-operator. Having said that I am not sure this should still be a requirement for GA. From my side it's fine to proceed to GA without this feature.

**Comment by weyert** [comment]
Yes, I think able to release a GA would make it available for a larger audience :)
